### PR TITLE
Delayed sampling, testing framework, and verification of the concept paper results.

### PR DESCRIPTION
**Pull request by vsenderov**
# Updated description of the PR

1. Support for manual delayed sampling.
    a. Several new distributions to accommodate delayed sampling.
        - two new files `delayed.cuh` and `delayed.cu` that have special distributions dedicated to delayed sampling;
        - some extensions to the regular distributions: `lomax`, `negativeBinomial`, `chiSquared` and `student_t`; for `chiSquared` a specialized implementation for the GPU and CPU was attempted, but, in the end, only left the universal code; attempt can be supplied if needed.
     b. Some math functions: `lchoose`, `log`, `logBeta`.

2. Existing phylogenetic models simplified to contain only a simple CRBD implementation and the generic stack implementation of Joey from before. The advanced phylogenetics models will be organized separately.

3. Unit testing framework. To run the tests you need to do ./unit_tests.sh from RootPPL base. You can play with the test parameters inside the script.

4. NaN weight workaround by Joey.

### CODE REVIEW

File: rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu [hunk 1]
```diff
@@ -0,0 +1,339 @@
+/*
+ * File clads2.cu defines the ClaDS2 model
```
Comment (david-broman) [inline rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu:2]: There are several files under a folder `junk`. Should they really be committed? If so, what is the purpose of the `junk` folder?

Comment (vsenderov) [reply rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu:2]: I briefly touched upon this in my PR description. This is one of the reasons why my initial idea was to discuss with Joey, before making the final PR. In a nutshell this is either older reference code, that I felt uneasy to delete, or some promising attempts, that have not been fully tested. They should go in a different branch, I suppose.

Comment (vsenderov) [reply rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu:2]: Comments on how to handle it, appreciated.

Comment (JoeyOhman) [reply rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu:2]: Many of these files are my old files. There are only two of these I think should stay:

1. `crbd/junk/condbd.cu`
2. `crbd/junk/stack.cuh`

These could turn out to be useful for reference when Daniel want to generate RootPPL code. However, they should eventually be removed as well. A directory for them, for now could be `crbd/generic_stack/`? Feel free to remove the rest of my old phylogenetics reference files that are no longer used. 

I do not know what to do with your files in the `junk` directories.

**Comment by JoeyOhman** [comment]
- Many new files are named with `kebab-case`. But the current file naming convention in RootPPL is `snake_case`. (Note that directories are named with `kebab-case` though)

- Is there a dependency on `R` for the unit tests? Perhaps it does not matter, but I had trouble executing them.

**Comment by vsenderov** [comment]
@david-broman 

I cleaned up my PR, removed the junk folders, changed the filename case as it should be as per yours and Joey's suggestions.

The number of phylogenetic models has also been decreased. Some of the working ideas have been moved away. Now we have only tested models.

- CRBD with immediate sampling
- CRBD with delayed sampling (survivorship bias not implemented yet)
- ClaDS2 with immediate sampling (not saving intermediate states)
- ClaDS2 (factorized) with immediate sampling (not saving intermediate states)
- ClaDS2 (factorized) with delayed sampling (saving intermediate states)

I also removed the verification of the concept paper. This should go in a separate repo as it is not an integral part of RootPPL.

File: rootppl/dists/dists_cpu.cuh [hunk 2]
```diff
@@ -60,4 +60,21 @@ unsigned int poisson(double lambda) {
...
+// DEV floating_t chi_squared(RAND_STATE_DECLARE floating_t k) {
+//   assert(0.0 < k);
+//   std::chi_squared_distribution<double> dist(k);
+
+//   #ifdef _OPENMP
+//   return dist(genWrappers[omp_get_thread_num()].gen);
+//   #else
+//   return dist(gen);
+//   #endif
+// }
```
Comment (dlunde) [inline rootppl/dists/dists_cpu.cuh:77]: Should this commented out code be removed?

File: rootppl/dists/dists_gpu.cuh [hunk 3]
```diff
@@ -84,4 +84,8 @@ __device__ floating_t gamma(curandState* randState, floating_t k, floating_t the
 }
 */
 
-#endif
\ No newline at end of file
+
+
+
+
+#endif
```
Comment (dlunde) [inline rootppl/dists/dists_gpu.cuh:91]: Restore original white space here (no actual change to this file)

File: rootppl/do_birds.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+fi
+# test var needs to be set 
+for model in $(ls $BIRDDIR)
+do
+    make clean
+    make model=$BIRDDIR/$model -j$CORES
+    ./program $NUMPART $NUMRUN
+    cp log_norm_const.txt $RESULTDIR/$model.logz
+    mv results/EXP-* $RESULTDIR/
+done
```
Comment (dlunde) [inline rootppl/do_birds.sh:19]: Should this file really be in the top directory? To me, it feels like this should go under `models`. Or, maybe we should even consider moving the phylogenetic stuff to it's own repository.

File: rootppl/log_norm_const.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+[-139.844]
```
Comment (dlunde) [inline rootppl/log_norm_const.txt:1]: This file should probably be deleted?

**Review by dlunde — COMMENTED** [review]
I think this can now be merged after some minor changes.

**Comment by vsenderov** [comment]
> * Many new files are named with `kebab-case`. But the current file naming convention in RootPPL is `snake_case`. (Note that directories are named with `kebab-case` though)
> * Is there a dependency on `R` for the unit tests? Perhaps it does not matter, but I had trouble executing them.

I think I mostly deleted kebab-case-named-files. However, still some camelCaseFiles remain in the tests directory. Let me know if still problem.

File: rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu (hunk 1 above)
Comment (vsenderov) [reply rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu]: We refactored those.

File: rootppl/dists/dists_cpu.cuh (hunk 2 above)
Comment (vsenderov) [reply rootppl/dists/dists_cpu.cuh:77]: done

File: rootppl/dists/dists_gpu.cuh (hunk 3 above)
Comment (vsenderov) [reply rootppl/dists/dists_gpu.cuh:91]: done

File: rootppl/do_birds.sh (hunk 4 above)
Comment (vsenderov) [reply rootppl/do_birds.sh:19]: this is removed now

File: rootppl/log_norm_const.txt (hunk 5 above)
Comment (vsenderov) [reply rootppl/log_norm_const.txt:1]: done

**Comment by vsenderov** [comment]
I addressed all issues in the comments.

In addition to that I back-ported the new CRBD version with "Jan's walk" and changed the default output of `smc.cu` to be the normalizing const without the verbosity (just a number).

**Comment by dlunde** [comment]
I believe this can be merged now. @vsenderov I'll present it on the Miking meeting today

File: rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu (hunk 1 above)
Comment (david-broman) [reply rootppl/models/phylogenetics/clads2/junk/attempt-clads2.cu]: We need to rename the folder and add a good description of what these files are. I do not like to add folders with 'junk' in the repo.

**Comment by david-broman** [comment]
The main description should probably be updated. It still says that there are junk folders.

Also, how are the test models organized now? Are there duplications with the benchmark?

**Comment by vsenderov** [comment]
# Updated description of the PR

1. Support for manual delayed sampling.
    a. Several new distributions to accommodate delayed sampling.
        - two new files `delayed.cuh` and `delayed.cu` that have special distributions dedicated to delayed sampling;
        - some extensions to the regular distributions: `lomax`, `negativeBinomial`, `chiSquared` and `student_t`; for `chiSquared` a specialized implementation for the GPU and CPU was attempted, but, in the end, only left the universal code; attempt can be supplied if needed.
     b. Some math functions: `lchoose`, `log`, `logBeta`.

2. Existing phylogenetic models simplified to contain only a simple CRBD implementation and the generic stack implementation of Joey from before. The advanced phylogenetics models will be organized a git submodule of `miking-benchmarks`.

3. Unit testing framework. To run the tests you need to do ./unit_tests.sh from RootPPL base. You can play with the test parameters inside the script.

4. NaN weight workaround by Joey.

**Comment by dlunde** [comment]
@vsenderov Can you replace you original message with the above message instead? Easier to find that way, if we need to revisit this PR in the future.

**Comment by vsenderov** [comment]
> @vsenderov Can you replace you original message with the above message instead? Easier to find that way, if we need to revisit this PR in the future.

Like so?

**Comment by dlunde** [comment]
Yes, looks good!

**Pull request merged**